pkg/controllers/management/aks/aks_cluster_handler.go

```diff
@@ -0,0 +1,740 @@
+	if err != nil {
+		return cluster, err
+	}
+
+	// AKS cluster and node pool statuses are not always immediately updated. This cause the AKSConfig to
+	// stay in "active" for a few seconds, causing the cluster to go back to "active".
+	timeout := time.NewTimer(10 * time.Second)
+	for {
+		select {
+		case event := <-w.ResultChan():
+			aksClusterConfigDynamic = event.Object.(*unstructured.Unstructured)
+			status, _ := aksClusterConfigDynamic.Object["status"].(map[string]interface{})
+			if status["phase"] == "active" {
+				continue
+			}
+
+			// this enqueue is necessary to ensure that the controller is reentered with the updating phase
+			e.clusterEnqueueAfter(cluster.Name, enqueueTime)
+			return e.setUnknown(cluster, apimgmtv3.ClusterConditionUpdated, "")
+		case <-timeout.C:
+			cluster, err = e.recordAppliedSpec(cluster)
+			if err != nil {
+				return cluster, err
+			}
+			return cluster, nil
+		}
+	}
+}
+
+// recordCAAndAPIEndpoint reads the AKSClusterConfig's secret once available. The CA cert and API endpoint are then copied to the cluster status.
+func (e *aksOperatorController) recordCAAndAPIEndpoint(cluster *mgmtv3.Cluster) (*mgmtv3.Cluster, error) {
+	backoff := wait.Backoff{
+		Duration: 2 * time.Second,
+		Factor:   2,
+		Jitter:   0,
+		Steps:    6,
+		Cap:      20 * time.Second,
+	}
+
+	var caSecret *corev1.Secret
+	err := wait.ExponentialBackoff(backoff, func() (bool, error) {
+		var err error
+		caSecret, err = e.secretsCache.Get(namespace.GlobalNamespace, cluster.Name)
+		if err != nil {
+			if !errors.IsNotFound(err) {
+				return false, err
+			}
+			logrus.Infof("waiting for cluster [%s] data needed to generate service account token", cluster.Name)
+			return false, nil
+		}
+		return true, nil
+	})
+	if err != nil {
+		return cluster, fmt.Errorf("failed waiting for cluster [%s] secret: %s", cluster.Name, err)
+	}
+
+	apiEndpoint := string(caSecret.Data["endpoint"])
+	caCert := string(caSecret.Data["ca"])
+	if cluster.Status.APIEndpoint == apiEndpoint && cluster.Status.CACert == caCert {
+		return cluster, nil
+	}
+
+	var currentCluster *mgmtv3.Cluster
+	err = retry.RetryOnConflict(retry.DefaultRetry, func() error {
+		currentCluster, err = e.clusterClient.Get(cluster.Name, v1.GetOptions{})
+		if err != nil {
+			return err
+		}
+		currentCluster.Status.APIEndpoint = apiEndpoint
+		currentCluster.Status.CACert = base64.StdEncoding.EncodeToString(caSecret.Data["ca"])
+		currentCluster, err = e.clusterClient.Update(currentCluster)
+		return err
+	})
+
+	return currentCluster, err
+}
+
+// generateAndSetServiceAccount uses the API endpoint and CA cert to generate a service account token. The token is then copied to the cluster status.
+func (e *aksOperatorController) generateAndSetServiceAccount(cluster *mgmtv3.Cluster) (*mgmtv3.Cluster, error) {
+
+	restConfig, err := e.getKubeConfig(cluster)
```

Comment:
@mjura can you explain how `GetClusterKubeConfig` works and where it lives ? I don't see it.

@rmweir can you take a look at this area. This looks weird to me. In EKS we use AWS api to get a new token and build clientset with that, then make the new SA. This looks like we are building a new kubeconfig.

Reply:
I have implemented method which is fetching ready `kubeconfig` from AKS AccessProfile based on Azure API
https://github.com/mjura/aks-operator/blob/master/controller/aks-cluster-config-handler.go#L394

For eks in this place it was generating neAccessToken
https://github.com/rancher/rancher/blob/master/pkg/controllers/management/eks/eks_cluster_handler.go#L702

I would prefer to keep this if it is possible, it is much nicer

Reply:
Have you had a chance to investigate this?

Reply:
AKS has option for setting up private cluster, but I will re-think how to combine `generateAndSetServiceAccount` and `generateSATokenWithPublicAPI`

Reply:
Since we are just using the function to grab the kubeconfig now, do we no longer need to make the secret in aks operator: https://github.com/rancher/aks-operator/blob/cdf6a514c052276a5ef815afdc9ca0cef8bb263c/controller/aks-cluster-config-handler.go#L476?

Reply:
Yes, we have to still create secret in aks-operator, because of CA data, which is used afterwards buy cluster dialer